### DISCUSSION TITLE
Consensus Protocols Comparison

### DISCUSSION THREAD

**Discussion by sa8** [consensus-shipyard/consensuslab#117] [Reading group]
The table below compares the papers read until now. We first present our main takeaways.
As a reminder, the properties that we are interested in in our reviews are:
- scaling to ten thousands nodes.
- censorship resistance. This specifically applies to WindowPosts that should be included in the chain 30 minutes after a sector is challenged on-chain.
- relatively low finality. 
  - Ideally, checkpoints coming from subnets should be included in the chain within 5 minutes (this may however be too strong of a requirement).
  - It should not be feasible to rewrite the history for a length > `PreCommitChallengeDelay` (currently 150 blocks).
- fairness.
- simplicity.

We started the formalization of these requirements in this [overleaf (WIP)](https://www.overleaf.com/read/grnswdbpkwck).
## Takeaway

- Most of the longest-chain and longest-chain improvement protocols are quite simple and good but suffer from long finality, which is problematic in Filecoin due to storage requirements (point 1 and 2 listed above). This is also true for Expected Consensus. For longest-chain protocols, in the ideal case (i.e., using SSLE), the finality is around 350 blocks which is longer than the 150 blocks required by`PreCommitChallengeDelay`. 
  - Prism and Ledger combiner have better finality but are more complex and are only proven secure in the context of proof-of-work. They are not implemented anywhere right now (although a prototype of Prism exists). Ledger Combiner may not translate well in the context of proof-of-space.
- Algorand, Avalanche, Spacemesh and Ebb-and-flow all are variants of permissionless PBFT-style consensus. They provide better finality than longest-chain protocols although they are more complex and usually scale less.
Spacemesh seems very complex and the paper is under-specified on some parts. It is not clear how well it scales. This is similar for Avalanche, the paper is not peer-reviewed, under-specified and the results seem too good to be true but the protocol is much simpler than Algorand and Spacemesh.
Algorand is good but only for up to a 20% adversary. With a 33% adversary it is quite long and inefficient.

Our favorite candidates to look at right now are:
- Prism which has good finality but is more complex than LC and only proven secure in the proof-of-work context.
- Avalanche: good candidate but may be too good to be true.
- Algorand: good performance and provably-secure but inefficient with an adversary that has >20% of the power.

An alternative idea from @xuechao2 to research more: [mix Snowball with EC](https://github.com/protocol/ConsensusLab/discussions/116).

## Comparative Table
In the table, _liveness_ is how long we need to wait for an honest block to appear in the chain, i.e., censorship-resistance.
_Partial synchrony_ means that the protocol keeps its safety property in periods of asynchrony (i.e., when we lose liveness).
_Weak synchrony_ means that the protocol needs synchrony for both safety and liveness, but can recover from asynchronous periods.
_Good case_ corresponds to the case where there is no adversary and synchrony.

|  <div style="width:290px">Paper</div>  | Scalability &nbsp; &nbsp; | Liveness | Finality | Incentive-compatibility | Simplicity | Provable security | <div style="width:290px">Order Fairness</div>  | Adversarial bound | Synchronicity | In production |
| -------- | -------- | -------- | -------- | -------- |-------- |-------- |-------- |-------- |-------- | -------- |
| [Longest-chain (LC) ](https://github.com/protocol/ConsensusLab/discussions/112) protocols: Ouroboros family, Nakamoto-PoS, PoSAT   | Very scalable (1 broadcast per round)   |  not good: k (finality parameter). This is in the worst case (i.e., with non-negligible but small probability). On average this will be shorter.  | k (~500 blocks) 4 hours in Filecoin  | No. Selfish mining. | Very simple   | Yes | No  | Ouroboros: <50%  Nakamoto-PoS/PoSAT: < 42.7% with c=32 (chosen by [Chia](https://docs.chia.net/docs/03consensus/attacks_and_countermeasures/))| Weak synchrony (required for formal proof) | Yes: Ouroboros in Cardano; a variant of Nakamoto-PoS in Chia |
|[Algorand](https://github.com/protocol/ConsensusLab/discussions/114#discussion-4168594)|O(nN) communication n: size of committee, N: network size | good: $\ell$ blocks with probability $1-(1/3)^\ell$ | 1 block but block creation takes longer with bigger adversary, e.g. 1 minute with 20% adv.|NA (no proof, no attacks)| medium | Yes | No | Max: 33%. The bigger the adversary is, the longer it takes to create a block. To be performant, <20% adversary.| Weak synchrony for safety; synchrony for liveness | Yes |
|[Spacesmesh](https://github.com/protocol/ConsensusLab/discussions/110) | O(n^2) (using special diffusion properties) | Good (multiple blocks per epoch, cannot censor blocks) | 1.5 to 7 minutes | $\epsilon$-incentive compatible |Complex (Hare part) | Yes | No | <33 % |Weak synchrony | Testnet (currently down) |
|[Colordag](https://github.com/protocol/ConsensusLab/discussions/105#discussion-4159532) | Same as LC | Same as LC | Same as LC | $\epsilon$-sure Nash Equilibrium |Simple |Yes (but only for PoW)  | No |<50% | Synchrony| No |
|[Order-fair consensus](https://github.com/protocol/ConsensusLab/discussions/109#discussion-4163204) (modulo-m) | Same as LC | Same as LC | twice of LC (2k) | Selfish mining | Simple | Yes (but only for PoW)  | Yes |<33% | Synchrony| No, but may be a candidate for Chainlink 2.0 (not 100% sure) |
|[Order-fair consensus](https://github.com/protocol/ConsensusLab/discussions/109#discussion-4163204) (multi-chain) | Same as LC | Same as LC | same as LC | Selfish mining | Simple | Yes (but only for PoW)  | Yes |<18% | Synchrony| No |
|[Parallel-chain protocols](https://github.com/protocol/ConsensusLab/discussions/106#discussion-4159980): Prism/Ledger combiner | Same as LC | small constant number of (<10) blocks | honest transaction: small constant number of (<10) blocks; double-spend transaction: same as LC| NA (no proof, no attacks) | Medium |Yes (Prism only has proof for PoW; ledger combiner has proof for both PoW and PoS)  | No |<50% | Synchrony| No |
|[Ebb-and-flow](https://github.com/protocol/ConsensusLab/discussions/107#discussion-4160406) | No. Needs an additional permissioned committee besides miners| Same as LC/Prism (depending on the underlying sub-protocol) | Same as LC/Prism (depending on the underlying sub-protocol) | NA (no proof, no attacks) | Medium | Yes | No | <50% adversarial miner + <33% adversarial committee member | Partial synchrony | No. Eth 2.0 tries to propose their own ebb-and-flow protocol, but never makes it right |
| [Avalanche](https://github.com/protocol/ConsensusLab/discussions/111#discussion-4165874) (Snowball protocol) | O(kN) communication k: protocol parameter (~10), N: network size | a few seconds under good case; worst case analysis missing | a few seconds under good case; worst case analysis missing | NA (no proof, no attacks) | Simple | No (particularly liveness analysis is missing) | No | Suspicious claim: [80% parameterized](https://www.avax.network/) From the paper: <0.3, safe but very slow; f<O(\sqrt{n}), fast | Safe under exponentially distributed message delays; live under synchrony | Yes |
|Longest chain with SSLE | Not scalable because of leader election   |  not good but better than LC : k (finality parameter as for LC)   | k (~350 blocks) 3 hours in Filecoin  | Selfish mining | Very simple   | Yes | No  | Ouroboros: <50%  Nakamoto-PoS/PoSAT: < 42.7% with c=32 (chosen by [Chia](https://docs.chia.net/docs/03consensus/attacks_and_countermeasures/))| Weak synchrony (required for formal proof) |No (SSLE not in production) |

**Comment by aakoshh** [comment]
Mina also [uses](https://docs.minaprotocol.com/en/architecture/proof-of-stake) Ouroboros, but they are considering integrating BFT algorithms for lower latency.

**Comment by LefKok** [comment]
Given that all protocols will use libp2p as communication layer does it make sense to use O(n) notation?
Also libp2p does not have formal liveness gurantees from what I understand so we might want to make sure that safety is not reliant on the liveness of the communication layer otherwise eclipse attacks become fair-game

**Reply by sa8** [reply]
> Given that all protocols will use libp2p as communication layer does it make sense to use O(n) notation?

Not sure I follow you, why would it not make sense?

**Reply by adlrocha** [reply]
_(I am probably joining the party late, but just came across this discussion :) )_
> Given that all protocols will use libp2p as communication layer does it make sense to use O(n) notation?

I guess it depends. If we use point-to-point communication between peers instead of a pubsub-based broadcast layer then the notation still makes sense (weather we use libp2p or not).

> Also libp2p does not have formal liveness gurantees from what I understand so we might want to make sure that safety is not reliant on the liveness of the communication layer otherwise eclipse attacks become fair-game

This is a good point. Gossipsub was [practically designed and implemented](https://research.protocol.ai/blog/2020/gossipsub-an-attack-resilient-messaging-layer-protocol-for-public-blockchains/) to prevent several types of attacks including eclipse attacks, but to the best of my knowledge there is no formal spec/analysis of the protocol. If we end up choosing a consensus protocol that relies on the liveliness of the transport later to ensure its safety we should probably spend some time digging into the formal specification of gossipsub (IIRC there's been some work here, and a few follow-up papers have been published. We can look into that).

**Comment by sa8** [comment]
Update: 
- Avalanche has been ruled out because of a weak threat model (adversary is root(n)).
- An application of [Bobtail](https://github.com/protocol/ConsensusLab/discussions/152) to the proof-of-storage setting is being considered.

**Reply by hmoniz** [reply]
I hope you don't mind me commenting here, but I also want to point out that Avalanche assumes a rather weak adversary that does not have any control over the scheduling of messages between correct processes. Instead, it assumes that message delay follows the exponential distribution. IMO, this is a rather strong assumption that might be difficult to hold in practice and can be a vector for attacks that target safety specifically.

**Comment by hmoniz** [comment]
This is a great comparative table! In the spirit of my comment above, I suggest maybe adding a column describing the power of the adversary beyond the resiliency bound. These protocols (and others) make different assumptions about the adversary (e.g., static, adaptive, scheduling limitations like no "after-the-fact" removals, atomic key erasures, etc.) that can have important practical consequences.

**Reply by sa8** [reply]
thank you for the suggestion, yes I agree that's a good idea!
